src/IR.h

```diff
@@ -814,6 +814,30 @@ struct Prefetch : public StmtNode<Prefetch> {
     static const IRNodeType _node_type = IRNodeType::Prefetch;
 };
 
+/** Lock all the Store nodes in the body statement.
+ *  Typically the lock is implemented by an atomic operation
+ *  (e.g. atomic add or atomic compare-and-swap).
+ *  However, if necessary, the node can access a mutex buffer through
+ *  mutex_name and mutex_args, by lowering this node into
+ *  calls to acquire and release the lock. */
+struct Atomic : public StmtNode<Atomic> {
+    std::string producer_name;
+    std::string mutex_name; // empty string if not using mutex
+    std::vector<Expr> mutex_indices;
```

Comment:
I don't think this field should be necessary, which would simplify the insertion of the mutexes (you won't have to worry about multiple dimensions and storage flattening interaction). Instead you can just lock the mutexes at each same single-dimensional coordinate that is accessed by Store nodes inside the body of the atomic

Reply:
Done. I noticed that there could be another Producer inside the Atomic due to inlining. So I still have to do some string matching to check whether the Store nodes is from the same producer.

Reply:
I removed the string matching.